pkg/controller/populators/upload-populator.go

```diff
@@ -0,0 +1,178 @@
+		if k8serrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	dataSourceRef := pvc.Spec.DataSourceRef
+	if dataSourceRef == nil || !IsPVCDataSourceRefKind(pvc, cdiv1.UploadSourceRef) || dataSourceRef.Name == "" {
+		log.V(1).Info("reconciled unexpected pvc, ignoring")
+		// Ignore PVCs that aren't for this populator to handle
+		return reconcile.Result{}, nil
+	}
+	uploadSource, err := r.getUploadSource(pvc.Namespace, dataSourceRef.Name)
+	if uploadSource == nil {
+		// Wait until dataSourceRef exists
+		return reconcile.Result{}, err
+	}
+
+	return r.reconcileTargetPVC(pvc, uploadSource)
+}
+
+func (r *UploadPopulatorReconciler) getUploadSource(namespace, name string) (*cdiv1.UploadSource, error) {
+	uploadSource := &cdiv1.UploadSource{}
+	uploadSourceKey := types.NamespacedName{Namespace: namespace, Name: name}
+	if err := r.client.Get(context.TODO(), uploadSourceKey, uploadSource); err != nil {
+		// reconcile will be triggered once created
+		if k8serrors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return uploadSource, nil
+}
+
+func updateAnnotations(pvc *corev1.PersistentVolumeClaim, source client.Object) {
+	pvc.Annotations[cc.AnnUploadRequest] = ""
+	uploadSource := source.(*cdiv1.UploadSource)
+	pvc.Annotations[cc.AnnContentType] = cc.GetContentType(string(uploadSource.Spec.ContentType))
+}
+
+func (r *UploadPopulatorReconciler) reconcileTargetPVC(pvc *corev1.PersistentVolumeClaim, uploadSource *cdiv1.UploadSource) (reconcile.Result, error) {
+	var waitForFirstConsumer bool
+	var ready bool
+	var err error
+	if pvc.Spec.StorageClassName != nil {
+		ready, waitForFirstConsumer, err = r.handleStorageClass(pvc)
+		if !ready || err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Get the PVC'
+	pvcPrimeName := PVCPrimeName(pvc)
+	pvcPrime, err := r.getPVCPrime(pvcPrimeName, pvc.Namespace)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if cc.IsUnbound(pvc) {
+		// If PVC' doesn't exist yet, create it
+		if pvcPrime == nil {
+			pvcPrime, err = r.createPVCPrime(pvc, uploadSource, waitForFirstConsumer, updateAnnotations)
+			if err != nil {
+				r.recorder.Eventf(pvc, corev1.EventTypeWarning, errCreatingPVCPrime, err.Error())
+				return reconcile.Result{}, err
+			}
+		} else {
+			// Wait upload completes
+			phase, _ := pvcPrime.Annotations[cc.AnnPodPhase]
+			if phase != string(corev1.PodSucceeded) {
+				if phase == string(corev1.PodFailed) {
+					r.recorder.Eventf(pvcPrime, corev1.EventTypeWarning, errUploadFailed, fmt.Sprintf(messageUploadFailed, pvc.Name))
+				}
+				return reconcile.Result{}, nil
+			}
+
+			if err := r.rebindPV(pvc, pvcPrime); err != nil {
+				return reconcile.Result{}, err
+			}
+		}
+		return reconcile.Result{}, nil
+	}
+
+	// Wait for the bind controller to rebind the PV
+	if pvcPrime != nil {
+		if corev1.ClaimLost != pvcPrime.Status.Phase {
+			return reconcile.Result{RequeueAfter: 2 * time.Second}, err
```

Comment:
You are already returning on all err != nil flows, so I guess you can return nil here?